src/Microsoft.DotNet.XHarness.Android/AdbRunner.cs

```diff
@@ -298,19 +305,62 @@ public void PushFiles(string localDirectory, string devicePath, bool removeIfPre
             _log.LogDebug($"Copied {filesToCopy.Length} files");
         }
 
-        public (string StandardOutput, string StandardError, int ExitCode) RunApkInstrumentation(string apkName, TimeSpan timeout)
+        public Dictionary<string, string> GetAttachedDevicesAndArchitectures()
         {
-            return RunApkInstrumentation(apkName, "", new Dictionary<string, string>(), timeout);
+            Dictionary<string, string> devicesAndArchitectures = new Dictionary<string, string>();
+
+            var result = RunAdbCommand("devices -l");
+            string standardOutput = result.StandardOutput;
+
+            // If the adb server is dead, this can return 0 exit code and no output; retry a few times if so)
+            int retriesLeft = 5;
+            while (retriesLeft-- > 0 && result.ExitCode == (int)AdbExitCodes.SUCCESS && standardOutput == "")
+            {
+                result = RunAdbCommand("devices -l");

```

Comment:
Should there be a timeout? Or is there one inside `RunAdbCommand`?

Reply:
We might want to also add a timeout here just in case, similar to what we do in iOS, you never know.

Reply:
Yeah, true, we get a lot of indefinite hangs.

So one timeout in-between the attempts, one for the call itself

Reply:
there is a timeout already when not providing an arg, but I can lower it for this one as it should be faster than some.